File: solutions/fsharp/change/Change.fs
module Change

open System.Collections.Generic

let change coins target =
    let changes = Dictionary<int, list<int>>()
    let amounts = Queue<int>()
    amounts.Enqueue 0
    changes[0] <- []
    
    while amounts.Count > 0 do
        let amount = amounts.Dequeue 
        let change = changes.Item amount
        .
       

let findFewestCoins coins target =
    if target < 0 then None else change coins target

File: solutions/scala/change/src/main/scala/Change.scala
import scala.collection.{SortedSet, mutable}

object Change:
  def findFewestCoins(amount: Int, denominations: List[Int]): Option[List[Int]] =
    optimal(amount, SortedSet.from(denominations))

  private def optimal(amount: Int, denominations: SortedSet[Int]): Option[List[Int]] =
    val changes = mutable.HashMap[Int, List[Int]](0 -> Nil)
    val amounts = mutable.ArrayDeque[Int](0)

    while (amounts.nonEmpty)
      val head = amounts.removeHead()
      val value = changes.get(head)
      if head == amount then
        return value.map(_.reverse)

      denominations.foreach { coin =>
        val total = head + coin
        if total <= amount && !changes.contains(total) then
          amounts addOne total
          val change = coin :: value.getOrElse(Nil)
          changes.put(total, change)
      }
    None

  private def greedy(amount: Int, denominations: SortedSet[Int]): Option[List[Int]] =
    def change(amount: Int, result: List[Int]): Option[List[Int]] =
      if amount == 0 then Option(result)
      else denominations.maxBefore(amount + 1)
        .flatMap(coin => change(amount - coin, coin :: result))

    change(amount, Nil)
